src/features/chart/redux/actions.ts

```diff
@@ -0,0 +1,28 @@
+import { UPDATE_CHART_RANGE, UPDATE_CHART_DATA } from "./constants";
+import { ActionWithPayload } from "../../../store/actions";
+
+export type UpdateChartRangeAction = ActionWithPayload<
+  typeof UPDATE_CHART_RANGE,
+  string
+>;
+
+export const updateChartAction = (
+  chartRange: string
+): UpdateChartRangeAction => ({
+  type: UPDATE_CHART_RANGE,
+  payload: chartRange
+});
+
+export type UpdateChartDataAction = ActionWithPayload<
+  typeof UPDATE_CHART_DATA,
+  { date: Date; close: number }[] | null
```

Comment:
q: Maybe it'll be better to create an `interface` or `type` which contains your data structure.

You'll then be able to pass this down to your `Middleware.js` file, so your `socket.on` function will become:
`socket.on("chartData", (payload: { date: Date; close: number }[]) => {`